feat(filtering): filter by query

QZera opened the pull request:
Closes [5288652](https://app.radicle.xyz/nodes/seed.radicle.gr/rad:z2BdUVZFvHdxRfdtGJQdSH2kyXNM6/issues/5288652c8587b5579d33ce2461601ce5bfd8d178)

## Product-y Notes

> For @gsaslis and @maninak 

### Dimming the rest of the UI while the filter input is focused

Tried this, got the best results with opacity around 80% but even then it was a bit hit and miss between dark and light themes. Might be able to make it work with more tinkering but elected not to spend any more time on it for the time being.

### `Ctrl/Cmd+F` as the shortcut for the input

I didn't really like the idea of hijacking an existing browser shortcut. After some research I found `/`to be a pretty common shortcut for search so I went with that.

### Input width

Given that we want filtering qualifiers in the future (e.g. `creator:zac`), queries might get long. I'm thinking we might want to have the input span the remaining available header width. Thoughts?

### Hiding the keyboard shortcut advertisement during input focus

I'm on the fence about this. If you feel we should keep it visible while the input is focused, just let me know and I'll change it.

## Tech-y/Code-y Notes

> For @maninak 

### To dos

I've left a few Todos which I want to do (get it?), either before this gets merged or after in another PR depending on time. Either way I didn't want them to hold up the review.

### [Accessibility] Using the `output` element and `<input type="search" />`

#### `output` element

I couldn't really find any information or cases of it being used to indicate search results. So instead, I set `aria-live="polite"` in each columns' `ul` element.

#### `<input type="search" />`

This is implemented [differently across browsers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search#differences_between_search_and_text_types) which would compromise our UX. In addition, I couldn't find information indicating that `type="search"` has any accessibility impact. 

Instead, I used `type="text"` with an aria-label and wrapped it in a `form` with `role="search"` [as suggested by MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search#search_form_labels_and_accessibility)

QZera commented on types/tasks.ts:
````diff
@@ -24,3 +24,27 @@ export type Patch = RadiclePatch & {
+ * ['', 'The', ' quick brown fox jumps over the lazy ', 'dog.']
+ * ```
+ */
+export interface TaskHighlights {
````
Tried a couple of things before ending up with this.

---
One was similar but with indexes for the highlighted segments and in combination with the original string you could compute the highlights.
 (e.g. `**fo**o b**a**r` would become `[[0, 1], [5,5]]`)

This felt like an extra step of complexity compared to the current solution.

---

Another thing I tried was passing the query to each card and then having the card perform the matching and highlighting internally (the filtering was still done on the tasks store).

That felt like a lot of logic for the card, and also not very flexible (e.g. I doubt it would work if we added fuzzy search)

maninak reviewed (changes requested):
I'll try to review the code later today but until then a few UI/UX points:


0. submitting the form e.g. with enter key should have it's submit event prevented so as to not reload the page (`@submit.prevent`)
0. filtering sould search across all taks we have in mem, even the ones hidden (e.g. because they are older or otherwise)
0. input font-size should be same as other controls (another case of issues when escaping the UI lib and going for bespoke custom components)
0. input outline on focus should be thinner, same as when task kind dropdown is being click&held or tabbed into (again, see previous comment about not using the UI lib)
0. when focused the X inside the input should be as far away from the right edge as the filter icon is from its left edge
0. the keyboard shortcut indicator shouldn't be selectable e.g. when pressing ctrl+a (`user-select: none`)
0. when "all" kind is used the input's placeholder should not use the word "tasks" just "filter"
0. if/when we have "kind:issue" etc modifiers we can consider removing the task kind selector alltogether
0. input should flex its width trying to take more horizontal space if value is too big for the field to show without scrolling
0. pressing esc should unfocus (blur) the input (counteracts the press-key-to-focus functionality)
0. I think the the `aria-live="polite"` isn't related to using `<output>`. I think for this one `aria-live="assertive"` would be better, but TBH id' just not use it at all unless I test it against an actual screenreader. Beyond that, the ouput elem isn't very common but it does offer meta info connecting the form with what is shown below it, so we should replace the generic div containing the columns with that element regardless. It doesn't hurt and only adds.
0. the input type affects many things on a meta level and since that's what we're using it for, again like above, we should use `type="search"`. Just as an example the icon on some mobile keyboards changes depedning on the input type.
0. using `/` for focusing the filterbox is very aggressive for an app where we're highly likely to have mutations and other inputs later where the user may be typing this characher and would instead have the focus jump to the filterbox. That's exactly why in my specs I said it should only apply for this route and (should be removed when the route changes away from it, e.g. when seeing app options or task details). Also overloading native functionality like `Ctrl + F` is fine as long as we're providing a superset of the native functionality, which is the case for us here.

maninak commented:
The more I play around with it the more I feel like fuzzy search is needed, but at the same time I feel like there needs to be "strict" matching mode resulting in the current behaviour (which should be default and how to switch modes is something we need to think hard about, but later).

For example (in this current demo seeded with radicle-explorer tasks) as a user I want to find all tasks that relate to the keywords "e2e" and "test". If I type "e2e test" I get many results, but for "test e2e" I get nothing. Fuzzy would solve that (buy definition ignoring term order), but for some terms the results could prob be too noisy including false positives.

QZera commented:
In response to https://github.com/cytechmobile/radicle-planning-boards/pull/80#pullrequestreview-2149645286

- [x] 1. Done, good catch
- [ ] 2. I think we discussed in our catchup not to do that, let's discuss further on Zulip
- [ ] 3. That's why I've used the `aria-label` for this input, as demonstrated in [this example](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search#search_form_labels_and_accessibility) from MDN
- [x] 4. Updated the font-size
- [x] 5-6. Used the `ring-2` class as suggested  
- [x] 7. Went with "Clear"
- [x] 8. Done
- [x] 9. Done
- [x] 10. Done
- [ ] 11. Nothing to be done for now
- [x] 12. Done
- [ ] 13. Agreed. Will have a look at this on the "second review round"
- [x] 14. Done
- [x] 15. Done (also removed the `aria-live` I had added to the columns)
- [ ] 16. This requires more effort than it seems, will have a look at this on the "second review round"
- [x] 17. Done. However, I'm still not a fan of this
- [x] 18. Done
- [ ] 19. Tried to do a quick fix by adding overflow scroll, but had issues with the kind select overflowing vertically, will have a look at this on the "second review round"

maninak reviewed (approved):
In the interest of time before our deadline, as agreed, I'm deferring the code review for later.

I also consider feedback items (3), (16) and (11) resolved too (though I'm curious to eventually know why this approach was preferred).

As far as I can tell with a quick look everything else was addressed. 

From the UX perspective this is a joy to use, well done! :clap:
